### PR TITLE
Add OutOfDisk error and use it in go_modules ecosystem

### PR DESCRIPTION
**Pull request by jasonrudolph**
Similar to the introduction of the `OutOfMemory` error type in 61e79897c8c79ef0fb2e2c3c23ca8050fcf02754, this pull request introduces an `OutOfDisk` error type.

Incorporating new error types typically requires teaching each ecosystem to make use of the error type. Instead of one long-running pull request to update every ecosystem, this pull request incorporates the `OutOfDisk` error type into the go_modules ecosystem, and paves the way for future pull requests to add it to other ecosystems.

To date, I've seen two stack traces for out of disk errors in the go_modules ecosystem. One includes STDERR content like this:

    rsc.io/sampler imports
    write /opt/go/gopath/pkg/mod/cache/vcs/0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef/info/attributes: no space left on device

And the other includes STDERR content like this:

    rsc.io/sampler imports
    golang.org/x/text/language: write /tmp/go-codehost-014108053: input/output error

So, when a subprocess errors out, the changes in this pull request will look for those sorts of patterns in STDERR and will raise a `Dependabot::OutOfDisk` error.

Prior to this change, we would instead raise a generic `Dependabot::DependabotError` in this situation. By raising a more specific error type, we can more easily help developers understand exactly why an update failed, and we can instrument this error type to track how frequently it's occurring.

### CODE REVIEW

File: go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb [hunk 1]
```diff
@@ -35,6 +35,11 @@ class GoModUpdater
           /go: ([^@\s]+)(?:@[^\s]+)?: .* declares its path as: ([\S]*)/m
         ].freeze
 
+        OUT_OF_DISK_REGEXES = [
+          %r{input/output error}.freeze,
+          /no space left on device/.freeze
+        ].freeze
```
Comment (brrygrdn) [inline go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:41]: 👍🏻

**Review by brrygrdn — APPROVED** [review]
✨  I like the approach of applying it to one ecosystem to begin with 👍🏻

**Comment by jasonrudolph** [comment]
I ran through the following steps to test out this change.

1. Using the macOS Disk Utility app, I created a volume with a quota of 10 megabytes [[screenshot](https://user-images.githubusercontent.com/2988/101808180-01acfa00-3ae4-11eb-8dc2-991d623d7dd7.png)]. (I tried going as small as 1 megabyte, but that's apparently not supported.) For this example, I named the volume `sandbox-10m`.

2. In my local checkout of dependabot/dependabot-core, I edited the `bin/docker-dev-shell` script to mount that volume as the `/tmp` directory.

    ```diff
    @@ -87,6 +87,7 @@ echo "$(tput setaf 2)=> running docker development shell$(tput sgr0)"
    CODE_DIR="/home/dependabot/dependabot-core"
    touch .core-bash_history
    docker run --rm -ti \
    +  -v "/Volumes/sandbox-10m:/tmp" \
      -v "$(pwd)/.core-bash_history:/home/dependabot/.bash_history" \
      -v "$(pwd)/.rubocop.yml:$CODE_DIR/.rubocop.yml" \
      -v "$(pwd)/bin:$CODE_DIR/bin" \
    ```

3. I then filled up most of the volume so there's very little space left. To do so, I copied this [file](https://github.com/dependabot/dependabot-core/files/5675038/filler.txt) into the volume. (There's almost surely a more elegant approach, but this did the trick.)

4. Run `bin/docker-dev-shell`

5. Then, I followed the steps in the README to perform a dry-run in the dev shell:

   ```
    $ cd omnibus && bundle install && cd - && bin/dry-run.rb go_modules rsc/quote

    Fetching gem metadata from https://rubygems.org/......
    Fetching gem metadata from https://rubygems.org/.
    Resolving dependencies...
    ...
    Bundle complete! 32 Gemfile dependencies, 104 gems now installed.
    Bundled gems are installed into `/home/dependabot/.bundle`
    /home/dependabot/dependabot-core
    "/home/dependabot/dependabot-core/d20201210-50-1gm4031"
    => cloning into /home/dependabot/dependabot-core/d20201210-50-1gzfu6x
    => parsing dependency files
    => updating 2 dependencies: rsc.io/quote/v3, rsc.io/sampler

    === rsc.io/quote/v3 (3.0.0)
    => checking for updates 1/2
    => latest available version is 3.1.0
    => latest allowed version is 3.1.0
    => requirements to unlock: own
    => requirements update strategy:
    => updating rsc.io/quote/v3 from 3.0.0 to 3.1.0
    Traceback (most recent call last):
      17: from bin/dry-run.rb:548:in `<main>'
      16: from bin/dry-run.rb:548:in `each'
      15: from bin/dry-run.rb:635:in `block in <main>'
      14: from bin/dry-run.rb:535:in `generate_dependency_files_for'
      13: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater.rb:34:in `updated_dependency_files'
      12: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:49:in `updated_go_mod_content'
      11: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:62:in `updated_files'
      10: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:66:in `update_files'
      9: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:196:in `in_repo_path'
      8: from /home/dependabot/dependabot-core/common/lib/dependabot/shared_helpers.rb:49:in `in_a_temporary_repo_directory'
      7: from /home/dependabot/dependabot-core/common/lib/dependabot/shared_helpers.rb:49:in `chdir'
      6: from /home/dependabot/dependabot-core/common/lib/dependabot/shared_helpers.rb:49:in `block in in_a_temporary_repo_directory'
      5: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:197:in `block in in_repo_path'
      4: from /home/dependabot/dependabot-core/common/lib/dependabot/shared_helpers.rb:170:in `with_git_configured'
      3: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:198:in `block (2 levels) in in_repo_path'
      2: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:86:in `block in update_files'
      1: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:123:in `run_go_mod_tidy'
    /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:264:in `handle_subprocess_error': go: downloading golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c (Dependabot::DependabotError)
    rsc.io/quote imports
      rsc.io/quote/v3 imports
      rsc.io/sampler imports
      golang.org/x/text/language: write /tmp/go-codehost-014108053: input/output error
    ```

    We can see above that we're getting the generic `Dependabot::DependabotError`.

6. After making the changes in this pull request, I ran the same steps again:

    ```
    $ cd omnibus && bundle install && cd - && bin/dry-run.rb go_modules rsc/quote
    ...
    => cloning into /tmp/d20201210-2311-kac2b5
    => parsing dependency files
    => updating 2 dependencies: rsc.io/quote/v3, rsc.io/sampler

    === rsc.io/quote/v3 (3.0.0)
    => checking for updates 1/2
    => latest available version is 3.1.0
    => latest allowed version is 3.1.0
    => requirements to unlock: own
    => requirements update strategy:
    => updating rsc.io/quote/v3 from 3.0.0 to 3.1.0
    Traceback (most recent call last):
      17: from bin/dry-run.rb:548:in `<main>'
      16: from bin/dry-run.rb:548:in `each'
      15: from bin/dry-run.rb:635:in `block in <main>'
      14: from bin/dry-run.rb:535:in `generate_dependency_files_for'
      13: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater.rb:34:in `updated_dependency_files'
      12: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:54:in `updated_go_mod_content'
      11: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:67:in `updated_files'
      10: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:71:in `update_files'
      9: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:207:in `in_repo_path'
      8: from /home/dependabot/dependabot-core/common/lib/dependabot/shared_helpers.rb:49:in `in_a_temporary_repo_directory'
      7: from /home/dependabot/dependabot-core/common/lib/dependabot/shared_helpers.rb:49:in `chdir'
      6: from /home/dependabot/dependabot-core/common/lib/dependabot/shared_helpers.rb:49:in `block in in_a_temporary_repo_directory'
      5: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:208:in `block in in_repo_path'
      4: from /home/dependabot/dependabot-core/common/lib/dependabot/shared_helpers.rb:170:in `with_git_configured'
      3: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:209:in `block (2 levels) in in_repo_path'
      2: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:89:in `block in update_files'
      1: from /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:174:in `run_go_get'
        /home/dependabot/dependabot-core/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb:276:in `handle_subprocess_error': go: updating go.sum: write /go.sum: input/output error (Dependabot::OutOfDisk)
    ```

    This time around, we now get the `Dependabot::OutOfDisk` error.

**Comment by jasonrudolph** [comment]
Looks like I've got a few CI failures to work through. I'll dig into them shortly.

**Comment by jasonrudolph** [comment]
Note: Even with these changes, there will likely be scenarios where a go_modules update can experience an out-of-disk error that gets reported as a generic error. In other words, this is a start, but is surely not a silver bullet. 😇 

For example, on the off chance that we run out of disk when writing out the `go.mod` file [here](https://github.com/dependabot/dependabot-core/blob/63beb92/go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb#L309), we won't yet detect that as a first-class `OutOfDisk` error. I anticipate us evolving the code to detect more scenarios over time (hopefully by detecting these failures at a higher level in the stack :crossed_fingers:).

**Review by feelepxyz — APPROVED** [review]
LGTM! Yeah I like the approach of just starting with go 👍

File: go_modules/spec/dependabot/go_modules/file_updater/go_mod_updater_spec.rb [hunk 2]
```diff
@@ -479,4 +479,36 @@
...
+      let(:requirements) { previous_requirements }
+      let(:previous_requirements) { [] }
+
+      it "detects 'input/output error'" do
+        stderr = <<~ERROR
+          rsc.io/sampler imports
+          golang.org/x/text/language: write /tmp/go-codehost-014108053: input/output error
+        ERROR
+
+        expect { updater.send(:handle_subprocess_error, stderr) }.to raise_error(Dependabot::OutOfDisk) do |error|
```
Comment (jurre) [inline go_modules/spec/dependabot/go_modules/file_updater/go_mod_updater_spec.rb:497]: It's a bit of a bummer that we have to test a private method here to verify this works, in this scenario I think this is fine, but I think it surfaces an opportunity for improvement in our design:

What if we had some object that we inject into our updater that was responsible for running these native/shell helpers, then we could write tests like this:

```ruby
it "detects 'input/output' error" do
  stderr = <<~ERROR
    rsc.io/sampler imports
    golang.org/x/text/language: write /tmp/go-codehost-014108053: input/output error
  ERROR

  mock_go_mod_runner = instance_double(
    GoModRunner,
    run_go_get: ->() { raise SubprocessError.new(stderr: stderr) }
  )

  updater = build_updater(
    dependency: "rsc.io/quote",
    new_version: "v1.5.2",
    previous_version: "v1.4.0",
    runner: mock_go_mod_runner
  )

  expect(updater.updated_go_mod_content).to raise_error(Dependabot::OutOfDisk) do |error|
    # ...
  end
end
```

_Another_ idea that came to mind is, what if we had VCR but for shelling out? Imagine a world in which we could do this:

```ruby
ShellRecorder.configure do
  hook_into: :open3
end

it "detects 'input/output' error" do
    updater = build_updater(
    dependency: "rsc.io/quote",
    new_version: "v1.5.2",
    previous_version: "v1.4.0"
  )

  ShellRecorder.record_shell_command("go_get_input_output_error") do
    expect(updater.updated_go_mod_content).to raise_error(Dependabot::OutOfDisk) do |error|
      # ...
    end
  end
end
```

Where running this once with the configuration that you mentioned would store that shell command in a `yaml` file, just like VCR does for web requests, in `fixtures/shell_recorder/file_updater/go_mod_updater_spec/go_get_input_output_error.yaml`:

```yaml
shell_interactions:
- command:
  binary: go
  options:
  - get
  - -u
  stdout: ""
  stderr: |
    rsc.io/sampler imports
    golang.org/x/text/language: write /tmp/go-codehost-014108053: input/output error
  status: 1
```

Any time we run that `ShellRecorder.record_shell_command` block when the cassette exists, we stub calls to `Open3` with that data. Alternatively we could also have our own `Open3` wrapper that we stub calls to instead, if that helps.

Aaaanyways, this is just a pipe dream and a rough idea that I had last night that I wanted to share, the PR is great as is 👍

**Pull request merged**